packages/@lwc/engine-core/src/framework/modules/props.ts

```diff
@@ -4,53 +4,50 @@
  * SPDX-License-Identifier: MIT
  * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
  */
-import { assert, isUndefined, keys } from '@lwc/shared';
-import { setProperty, getProperty } from '../../renderer';
+import { assert, isNull, isUndefined, keys } from '@lwc/shared';
+
+import { getProperty, setProperty } from '../../renderer';
 import { VElement } from '../../3rdparty/snabbdom/types';
 
+import { EmptyObject } from '../utils';
+
 function isLiveBindingProp(sel: string, key: string): boolean {
     // For properties with live bindings, we read values from the DOM element
     // instead of relying on internally tracked values.
     return sel === 'input' && (key === 'value' || key === 'checked');
 }
 
-function update(oldVnode: VElement, vnode: VElement) {
-    const props = vnode.data.props;
-
+export function patchProps(oldVnode: VElement | null, vnode: VElement) {
+    const { props } = vnode.data;
     if (isUndefined(props)) {
         return;
     }
-    const oldProps = oldVnode.data.props;
+
+    const oldProps = isNull(oldVnode) ? EmptyObject : oldVnode.data.props;
     if (oldProps === props) {
         return;
     }
 
     if (process.env.NODE_ENV !== 'production') {
         assert.invariant(
-            isUndefined(oldProps) || keys(oldProps).join(',') === keys(props).join(','),
+            oldProps === EmptyObject || keys(oldProps).join(',') === keys(props).join(','),
             'vnode.data.props cannot change shape.'
         );
     }
```

Comment:
This condition will never be true as long as the VNodes are generated by the template compiler. I do recall we added this assertion when we were still investing in how to compile HTML to JavaScript. I don't think those checks are necessary anymore.

It's certainly something that I will remove in a follow up PR.